rfcs/0003-timestamps-and-ttl.md

```diff
@@ -0,0 +1,329 @@
+- [Implementation](#implementation)
+  * [Data Format](#data-format)
+  * [Compaction](#compaction)
+- [Rejected Alternatives](#rejected-alternatives--follow-ups)
+- [Updates](#updates)
+
+<!-- TOC end -->
+
+Status: In Discussion
+
+Authors:
+* [Almog Gavra](https://github.com/agavra)
+
+References:
+* https://github.com/slatedb/slatedb/issues/218
+* https://github.com/slatedb/slatedb/issues/225
+* https://github.com/facebook/rocksdb/wiki/User-defined-Timestamp
+* https://github.com/facebook/rocksdb/wiki/Time-to-Live
+* https://github.com/facebook/rocksdb/wiki/Compaction-Filter
+* https://dl.acm.org/doi/pdf/10.1145/3483840
+* https://disc-projects.bu.edu/lethe/
+* https://dl.acm.org/doi/10.1145/3555041.3589719
+
+## Motivation
+
+SlateDB does not have a native understanding of time in data operations (it does
+in regard to compaction/garbage collection, discussion on this later). This introduces
+additional complexity to operators that want to implement semantics that depend on the
+insertion/modification time of rows. The most common usages of timestamps are:
+
+1. **Time-To-Live**: many applications, either for compliance or space-saving purposes
+   require purging data a certain amount of time after its insertion. Use cases with time
+   series data and privacy requirements such as GDPR come to mind here, though ensuring that 
+   deleted data is collected within a time bound is left for a future exercise.
+2. **Resolving Out-of-Order Operations**: some applications may process incoming requests
+   in a manner that results in out-of-order insertions to SlateDB (stream processing applications
+   are frequently subject to this type of pattern). Using insertion timestamps to resolve which
+   version of a key is more recent may be preferable to using the clock at the time of insertion.
+3. **Historical Reads / MVCC**: while SlateDB provides snapshot capabilities to ensure reads are
+   consistent and repeatable, it requires a snapshot to have been taken at the desired time. 
+   increasing the number of snapshots adds significant storage overhead. Instead, it is possible
+   to trade read-latency for space amplification by maintaining multiple versions of keys at
+   different timestamps to allow for similar "historical" read operations.
+
+## Goals
+
+- Define the API for specifying timestamps via user defined clocks
+- Define the changes to the on-disk format of stored rows
+- Outline mechanism to support Time-To-Live (TTL) via Compaction Filters
+
+## Out of Scope
+
+While the following are out of scope for the initial implementation, the design should
+not preclude:
+
+- Handling out-of-order insertions is not in scope for this RFC, though we make sure the design
+  accounts for the future possibility (see rejected alternatives section)
+- Outlining compaction strategies that are optimized for TTL (such as Cassandra's
+  [TWCS](https://cassandra.apache.org/doc/stable/cassandra/operating/compaction/twcs.html)). This
```

Comment:
Even if the file sort order is not changed to support `TWCS`, isn't it still possible to apply a TTL at file level instead of key level, if the key is monotonically non decreasing? 
Still reading the doc. Is handling that out of scope for this design? Should we track it as a separate issue?

Reply:
I gave it some thought, what we could do is on file creation persist the "max TTL" if every row in the SST has a TTL, this would help us optimize expiring an entire SST if that situation exists -- the thing is that I feel it's pretty unlikely with compaction strategies other than TWCS because they're constantly being updated with new rows.

Reply:
One scenario that I can think of is when the key has timestamp prefix. Insert order would naturally be mostly timestamp ordered, and older keys (== older timestamps) that are out of TTL would be together.

Reply:
Yeah I think this optimization is worthwhile (though doesn't need to be included in this RFC). It's a basic requirement in the way SlateDB is implemented to day that the sorted runs have to be maintained in insertion order We can loosen this eventually if we want to support more flexible compaction, but we'll potentially need to search more runs when reading. So if the data is being inserted in timestamp order, then the deeper runs should have older data.